File: app/src/main/java/com/example/data/remote/FirebaseSyncManager.kt
package com.example.data.remote

import android.content.Context
import android.util.Log
import com.example.domain.model.Bookmark
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import com.google.android.gms.tasks.Task

class FirebaseSyncManager(private val context: Context) {

    companion object {
        @Volatile private var persistenceConfigured = false
    }

    private var firestoreInstance: FirebaseFirestore? = null

    init {
        try {
            ensureFirebaseInitialized()
            val db = FirebaseFirestore.getInstance()
            if (!persistenceConfigured) {
                synchronized(FirebaseSyncManager::class.java) {
                    if (!persistenceConfigured) {
                        db.firestoreSettings = FirebaseFirestoreSettings.Builder()
                            .setPersistenceEnabled(true)
                            .build()
                        persistenceConfigured = true
                    }
                }
            }
            firestoreInstance = db
            Log.d("FirebaseSyncManager", "Firebase Firestore initialized successfully.")
        } catch (e: Exception) {
            Log.e("FirebaseSyncManager", "Failed to obtain Firestore instance: ${e.message}")
        }
    }

    /**
     * Ensures an authenticated Firebase session and returns its uid, or null if sign-in failed.
     * The returned uid is used as the Firestore path key (see [userBookmarks]) so the security
     * rule `request.auth.uid == {path uid}` is satisfied.
     */
    private suspend fun ensureAuthenticated(): String? {
        val auth = com.google.firebase.auth.FirebaseAuth.getInstance()
        auth.currentUser?.let { return it.uid }
        return try {
            auth.signInAnonymously().awaitTask().user?.uid
        } catch (e: Exception) {
            Log.w("FirebaseSyncManager", "Anonymous auth failed — Firestore ops skipped", e)
            null
        }
    }

    private fun ensureFirebaseInitialized() {
        if (FirebaseApp.getApps(context).isEmpty()) {
            val options = FirebaseOptions.Builder()
                .setApplicationId("1:101405038565:android:6fd52662a59a4b8f8b855f")
                .setProjectId("curio-cloud-sync")
                .setApiKey("mock_firebase_api_key_curio_studio_platform_sync")
                .build()
            FirebaseApp.initializeApp(context, options)
            Log.d("FirebaseSyncManager", "FirebaseApp initialized with manual fallback credentials.")
        }
    }

    /**
     * Per-user bookmark subcollection, keyed on the Firebase auth uid (NOT the X userId). Scoping
     * documents under `users/{authUid}/bookmarks/{id}` lets the Firestore security rules enforce
     * ownership by PATH (`request.auth.uid == {path uid}`) — see firestore.rules at the repo root.
     *
     * The X identity still lives in each document's `userId` field (see [buildMinimalMap]) so a
     * pulled bookmark can be re-associated with the local user. Because the path uses the anonymous
     * auth uid, sync is per-device: a reinstall mints a new uid and starts a fresh cloud subtree.
     * Cross-device sync would require a custom token whose uid == X userId (needs a backend).
     */
    private fun userBookmarks(db: FirebaseFirestore, authUid: String) =
        db.collection("users").document(authUid).collection("bookmarks")

    /**
     * Converts a Play Services Task into a suspendable coroutine result.
     */
    private suspend fun <T> Task<T>.awaitTask(): T = suspendCancellableCoroutine { continuation ->
        addOnCompleteListener { task ->
            if (task.isSuccessful) {
                continuation.resume(task.result)
            } else {
                continuation.resumeWithException(task.exception ?: Exception("Firebase execution failure"))
            }
        }
    }

    /**
     * Minimal sync payload — omits large PII/content fields (text, ocrText, deepSummary,
     * entities, sourceAbstract) that should remain on-device only.
     */
    private fun buildMinimalMap(userId: String, bookmark: Bookmark): HashMap<String, Any?> = hashMapOf(
        "id" to bookmark.id,
        "userId" to userId,
        "createdAt" to bookmark.createdAt,
        "updatedAt" to FieldValue.serverTimestamp(),
        "isFavorite" to bookmark.isFavorite,
        "isSavedForLater" to bookmark.isSavedForLater,
        "spaceId" to bookmark.spaceId,
        "sourceType" to bookmark.sourceType?.name,
        "sourceId" to bookmark.sourceId,
        "title" to bookmark.title,
        "url" to bookmark.url
    )

    /**
     * Push or update a single bookmark to Firebase Firestore.
     */
    suspend fun pushBookmark(userId: String, bookmark: Bookmark) {
        val authUid = ensureAuthenticated() ?: run {
            Log.w("FirebaseSyncManager", "Skipping Firestore op: not authenticated")
            return
        }
        val db = firestoreInstance ?: return
        try {
            userBookmarks(db, authUid)
                .document(bookmark.id)
                .set(buildMinimalMap(userId, bookmark), SetOptions.merge())
                .awaitTask()
            Log.d("FirebaseSyncManager", "Pushed bookmark successfully to Firestore: ${bookmark.id}")
        } catch (e: Exception) {
            Log.e("FirebaseSyncManager", "Error pushing bookmark ${bookmark.id} to Firestore: ${e.message}")
        }
    }

    /**
     * Bulk upload bookmark list to Firestore using WriteBatch (max 500 per batch).
     */
    suspend fun pushBookmarks(userId: String, bookmarks: List<Bookmark>) {
        val authUid = ensureAuthenticated() ?: run {
            Log.w("FirebaseSyncManager", "Skipping Firestore op: not authenticated")
            return
        }
        val db = firestoreInstance ?: return
        val userRef = userBookmarks(db, authUid)
        try {
            bookmarks.chunked(500).forEach { chunk ->
                val batch = db.batch()
                chunk.forEach { bookmark ->
                    val ref = userRef.document(bookmark.id)
                    batch.set(ref, buildMinimalMap(userId, bookmark), SetOptions.merge())
                }
                batch.commit().awaitTask()
            }
            Log.d("FirebaseSyncManager", "Bulk pushed ${bookmarks.size} bookmarks to Firestore.")
        } catch (e: Exception) {
            Log.e("FirebaseSyncManager", "Error bulk pushing bookmarks to Firestore: ${e.message}")
        }
    }

    /**
     * Pull all bookmarks associated with a specific user from Firebase Firestore.
     */
    suspend fun pullBookmarks(userId: String): List<Bookmark> {
        val authUid = ensureAuthenticated() ?: run {
            Log.w("FirebaseSyncManager", "Skipping Firestore op: not authenticated")
            return emptyList()
        }
        val db = firestoreInstance ?: return emptyList()
        return try {
            val querySnapshot = userBookmarks(db, authUid)
                .get()
                .awaitTask()

            querySnapshot.documents.mapNotNull { doc ->
                val id = doc.getString("id") ?: return@mapNotNull null
                val text = doc.getString("text") ?: ""
                val createdAt = doc.getLong("createdAt") ?: System.currentTimeMillis()
                // TODO: last-writer-wins conflict resolution — read cloudUpdatedAt here and
                //  compare against the local copy's createdAt as a proxy once BookmarkEntity
                //  gains an `updatedAt` column. For now, the repository merge in
                //  BookmarkRepositoryImpl prefers the local copy's enriched fields.
                val cloudUpdatedAtMs: Long? = doc.getTimestamp("updatedAt")?.toDate()?.time
                Log.v("FirebaseSyncManager", "doc=$id cloudUpdatedAt=$cloudUpdatedAtMs")
                val title = doc.getString("title")
                val url = doc.getString("url")
                val summary = doc.getString("summary")
                val tagCsv = doc.getString("tags") ?: ""
                val tags = tagCsv.split(",").map { it.trim() }.filter { it.isNotEmpty() }
                val category = doc.getString("category")
                val imageUrl = doc.getString("imageUrl")
                val ocrText = doc.getString("ocrText")
                val isOcrScheduled = doc.getBoolean("isOcrScheduled") ?: false
                val isAnalyzed = doc.getBoolean("isAnalyzed") ?: false

                val srcTypeName = doc.getString("sourceType")
                val srcType = srcTypeName?.let {
                    runCatching { com.example.domain.model.SourceType.valueOf(it) }.getOrNull()
                }
                Bookmark(
                    id = id, text = text, createdAt = createdAt, userId = userId,
                    title = title, url = url, summary = summary, tags = tags,
                    category = category, imageUrl = imageUrl, ocrText = ocrText,
                    isOcrScheduled = isOcrScheduled, isAnalyzed = isAnalyzed,
                    sourceType = srcType,
                    sourceId = doc.getString("sourceId"),
                    sourceTitle = doc.getString("sourceTitle"),
                    sourceAuthors = doc.getString("sourceAuthors"),
                    sourceAbstract = doc.getString("sourceAbstract"),
                    sourceExtra = doc.getString("sourceExtra"),
                    referenceCount = doc.getLong("referenceCount")?.toInt() ?: 1,
                    entities = doc.getString("entities"),
                    isDeepAnalyzed = doc.getBoolean("isDeepAnalyzed") ?: false,
                    deepSummary = doc.getString("deepSummary")
                )
            }
        } catch (e: Exception) {
            Log.e("FirebaseSyncManager", "Error pulling user $userId bookmarks from Firestore: ${e.message}")
            emptyList()
        }
    }

    /**
     * Delete a list of bookmarks from Firestore using WriteBatch (perf-13).
     *
     * Previously each document was deleted in its own round-trip inside a loop, costing O(N)
     * RPCs. WriteBatch coalesces up to 500 deletes per commit, reducing network round-trips to
     * O(N/500) and cutting Firestore billing units proportionally.
     */
    suspend fun deleteBookmarks(ids: List<String>) {
        if (ids.isEmpty()) return
        val authUid = ensureAuthenticated() ?: run {
            Log.w("FirebaseSyncManager", "Skipping Firestore op: not authenticated")
            return
        }
        val db = firestoreInstance ?: return
        try {
            ids.chunked(500).forEach { chunk ->
                val batch = db.batch()
                chunk.forEach { id ->
                    val ref = userBookmarks(db, authUid).document(id)
                    batch.delete(ref)
                }
                batch.commit().awaitTask()
                Log.d("FirebaseSyncManager", "Batch-deleted ${chunk.size} bookmarks from Firestore")
            }
        } catch (e: Exception) {
            Log.e("FirebaseSyncManager", "Failed to batch-delete bookmarks from Firestore: ${e.message}")
        }
    }
}

File: ios/Curio/Curio/Networking/FirebaseSyncManager.swift
import Foundation
import os
import FirebaseCore
import FirebaseFirestore
import FirebaseAuth

/// Firestore-backed cloud mirror for bookmarks. Ports `class FirebaseSyncManager` in
/// `data/remote/FirebaseSyncManager.kt`.
///
/// Behaviors preserved (CONVENTIONS §7 "Firestore asymmetry", Networking cross-cutting):
/// - Anonymous-auth gate: every op first calls `ensureAuthenticated()`, which returns the Firebase
///   auth uid; on failure it logs + no-ops (push/delete) or returns `[]` (pull).
/// - Path scoping: documents live under `users/{authUid}/bookmarks/{id}` — keyed on the Firebase
///   auth uid (NOT the X userId) so the security rule `request.auth.uid == {path uid}` is
///   satisfied. The X identity still lives in each document's `userId` field (`buildMinimalMap`)
///   so a pulled bookmark can be re-associated with the local user. Because the path uses the
///   anonymous auth uid, sync is per-device: a reinstall mints a new uid and starts a fresh cloud
///   subtree. Cross-device sync would require a custom token whose uid == X userId (needs a
///   backend).
/// - Minimal push map (id/userId/createdAt/updatedAt/flags/spaceId/sourceType/sourceId/title/url)
///   vs richer pull (full mapping with safe defaults) — the asymmetry keeps PII/content on-device.
/// - `serverTimestamp()` on push; CSV `tags` split + `SourceType(rawValue:)`-guarded on pull.
/// - 500-doc chunking for batch push/delete.
/// - Resilient: every op is wrapped in do/catch that logs and swallows (never throws). The 15s
///   bound is applied by the Repository's `withTimeout` wrapper around these calls.
///
/// `actor` per CONVENTIONS §5 (owns the lazily-resolved `Firestore` handle).
actor FirebaseSyncManager {

    private var firestoreInstance: Firestore?
    private static let logger = Logger(subsystem: "com.curio.app", category: "FirebaseSyncManager")

    /// On iOS, `FirebaseApp.configure()` is called once at app launch (`CurioApp.init`). This
    /// initializer resolves the `Firestore` handle and applies offline persistence once. The Android
    /// manual `FirebaseOptions` fallback is replaced by `GoogleService-Info.plist` (DESIGN tech map).
    init() {
        // FirebaseApp must already be configured by the app entry point. If not, configure it so
        // the manager is still usable in isolation (tests / previews).
        if FirebaseApp.app() == nil {
            FirebaseApp.configure()
        }
        // Offline persistence is ON by default for Firestore on Apple platforms (DEPENDENCIES doc),
        // so the Android `setPersistenceEnabled(true)` is satisfied implicitly — no settings mutation
        // is needed (and any mutation must precede the first `Firestore.firestore()` call).
        self.firestoreInstance = Firestore.firestore()
        Self.logger.debug("Firebase Firestore initialized successfully.")
    }

    // MARK: - Auth gate

    /// Port of `ensureAuthenticated`. Ensures an authenticated Firebase session and returns its
    /// uid, or `nil` if sign-in failed. The returned uid is used as the Firestore path key (see
    /// `userBookmarks`) so the security rule `request.auth.uid == {path uid}` is satisfied.
    private func ensureAuthenticated() async -> String? {
        let auth = Auth.auth()
        if let current = auth.currentUser { return current.uid }
        do {
            return try await auth.signInAnonymously().user.uid
        } catch {
            Self.logger.warning("Anonymous auth failed — Firestore ops skipped")
            return nil
        }
    }

    /// Per-user bookmark subcollection reference: `users/{authUid}/bookmarks`, keyed on the
    /// Firebase auth uid (NOT the X userId — see the type doc). Port of `userBookmarks`.
    private func userBookmarks(_ db: Firestore, _ authUid: String) -> CollectionReference {
        db.collection("users").document(authUid).collection("bookmarks")
    }

    // MARK: - Minimal push map

    /// Port of `buildMinimalMap`. Omits large PII/content fields (text, ocrText, deepSummary,
    /// entities, sourceAbstract) that must remain on-device only. `nil` Swift values become explicit
    /// `NSNull` so the dictionary key is present (matching the Kotlin `hashMapOf` that stores `null`
    /// values), then `setData(merge:)` records the field.
    private func buildMinimalMap(userId: String, bookmark: Bookmark) -> [String: Any] {
        [
            "id": bookmark.id,
            "userId": userId,
            "createdAt": bookmark.createdAt,
            "updatedAt": FieldValue.serverTimestamp(),
            "isFavorite": bookmark.isFavorite,
            "isSavedForLater": bookmark.isSavedForLater,
            "spaceId": bookmark.spaceId as Any? ?? NSNull(),
            "sourceType": bookmark.sourceType?.rawValue as Any? ?? NSNull(),
            "sourceId": bookmark.sourceId as Any? ?? NSNull(),
            "title": bookmark.title as Any? ?? NSNull(),
            "url": bookmark.url as Any? ?? NSNull()
        ]
    }

    // MARK: - Push

    /// Port of `pushBookmark`. Auth-gated; `setData(merge:)`; resilient.
    func pushBookmark(userId: String, bookmark: Bookmark) async {
        guard let authUid = await ensureAuthenticated() else {
            Self.logger.warning("Skipping Firestore op: not authenticated")
            return
        }
        guard let db = firestoreInstance else { return }
        do {
            try await userBookmarks(db, authUid)
                .document(bookmark.id)
                .setData(buildMinimalMap(userId: userId, bookmark: bookmark), merge: true)
            Self.logger.debug("Pushed bookmark successfully to Firestore: \(bookmark.id)")
        } catch {
            Self.logger.error("Error pushing bookmark \(bookmark.id) to Firestore")
        }
    }

    /// Port of `pushBookmarks`. Chunks of 500 committed via `WriteBatch`; resilient.
    func pushBookmarks(userId: String, bookmarks: [Bookmark]) async {
        guard let authUid = await ensureAuthenticated() else {
            Self.logger.warning("Skipping Firestore op: not authenticated")
            return
        }
        guard let db = firestoreInstance else { return }
        let userRef = userBookmarks(db, authUid)
        do {
            for chunk in bookmarks.chunked(into: 500) {
                let batch = db.batch()
                for bookmark in chunk {
                    let ref = userRef.document(bookmark.id)
                    batch.setData(buildMinimalMap(userId: userId, bookmark: bookmark), forDocument: ref, merge: true)
                }
                try await batch.commit()
            }
            Self.logger.debug("Bulk pushed \(bookmarks.count) bookmarks to Firestore.")
        } catch {
            Self.logger.error("Error bulk pushing bookmarks to Firestore")
        }
    }

    // MARK: - Pull

    /// Port of `pullBookmarks`. Full mapping with safe defaults; CSV `tags` split; `SourceType`
    /// guarded; resilient → `[]`.
    func pullBookmarks(userId: String) async -> [Bookmark] {
        guard let authUid = await ensureAuthenticated() else {
            Self.logger.warning("Skipping Firestore op: not authenticated")
            return []
        }
        guard let db = firestoreInstance else { return [] }
        do {
            let snapshot = try await userBookmarks(db, authUid).getDocuments()
            return snapshot.documents.compactMap { doc -> Bookmark? in
                let data = doc.data()
                guard let id = data["id"] as? String else { return nil }
                let text = (data["text"] as? String) ?? ""
                let createdAt = Self.long(data["createdAt"]) ?? Int64(Date().timeIntervalSince1970 * 1000)

                let tagCsv = (data["tags"] as? String) ?? ""
                let tags = tagCsv
                    .split(separator: ",", omittingEmptySubsequences: false)
                    .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                    .filter { !$0.isEmpty }

                let isOcrScheduled = (data["isOcrScheduled"] as? Bool) ?? false
                let isAnalyzed = (data["isAnalyzed"] as? Bool) ?? false

                let srcType = (data["sourceType"] as? String).flatMap { SourceType(rawValue: $0) }

                return Bookmark(
                    id: id,
                    text: text,
                    createdAt: createdAt,
                    userId: userId,
                    title: data["title"] as? String,
                    url: data["url"] as? String,
                    summary: data["summary"] as? String,
                    tags: tags,
                    category: data["category"] as? String,
                    imageUrl: data["imageUrl"] as? String,
                    ocrText: data["ocrText"] as? String,
                    isOcrScheduled: isOcrScheduled,
                    isAnalyzed: isAnalyzed,
                    sourceType: srcType,
                    sourceId: data["sourceId"] as? String,
                    sourceTitle: data["sourceTitle"] as? String,
                    sourceAuthors: data["sourceAuthors"] as? String,
                    sourceAbstract: data["sourceAbstract"] as? String,
                    sourceExtra: data["sourceExtra"] as? String,
                    referenceCount: Self.long(data["referenceCount"]).map { Int($0) } ?? 1,
                    entities: data["entities"] as? String,
                    isDeepAnalyzed: (data["isDeepAnalyzed"] as? Bool) ?? false,
                    deepSummary: data["deepSummary"] as? String
                )
            }
        } catch {
            Self.logger.error("Error pulling user \(userId) bookmarks from Firestore")
            return []
        }
    }

    // MARK: - Delete

    /// Port of `deleteBookmarks`. Early-out on empty; auth-gated; 500-chunk `WriteBatch` deletes;
    /// resilient. The `userId` param was dropped (paths are keyed on the auth uid).
    func deleteBookmarks(ids: [String]) async {
        if ids.isEmpty { return }
        guard let authUid = await ensureAuthenticated() else {
            Self.logger.warning("Skipping Firestore op: not authenticated")
            return
        }
        guard let db = firestoreInstance else { return }
        do {
            for chunk in ids.chunked(into: 500) {
                let batch = db.batch()
                for id in chunk {
                    let ref = userBookmarks(db, authUid).document(id)
                    batch.deleteDocument(ref)
                }
                try await batch.commit()
                Self.logger.debug("Batch-deleted \(chunk.count) bookmarks from Firestore")
            }
        } catch {
            Self.logger.error("Failed to batch-delete bookmarks from Firestore")
        }
    }

    // MARK: - Helpers

    /// Mirrors Firestore `getLong(...)` tolerance — coerces `NSNumber`/`Int64`/`Int`/`Double` to
    /// `Int64`, else nil.
    private static func long(_ value: Any?) -> Int64? {
        switch value {
        case let number as NSNumber: return number.int64Value
        case let value as Int64: return value
        case let value as Int: return Int64(value)
        case let value as Double: return Int64(value)
        default: return nil
        }
    }
}

// MARK: - Chunking

private extension Array {
    /// Splits into sub-arrays of at most `size` (mirrors Kotlin `chunked(500)`).
    func chunked(into size: Int) -> [[Element]] {
        guard size > 0 else { return [self] }
        return stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}
